pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,375 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+var globalStore *Store
+
+const (
+	retryCollectorInterval = 30 * time.Second
+	pullCollectorInterval  = 5 * time.Second
+	eventBundleChTimeout   = 1 * time.Second
+	eventChBufferSize      = 50
+)
+
+type subscriber struct {
+	name   string
+	ch     chan EventBundle
+	filter *Filter
+}
+
+// Store contains metadata for workloads.
+type Store struct {
+	storeMut sync.RWMutex
+	store    map[Kind]map[string]Entity
+
+	subscribersMut sync.RWMutex
+	subscribers    []subscriber
+
+	candidates map[string]Collector
+	collectors map[string]Collector
+
+	eventCh chan []Event
+}
+
+// NewStore creates a new workload metadata store. Call Run to start it.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
```

Comment:
This is worth including in the docs for this method, or maybe RegisterCollector: the factory is invoked to create a new collector for each store.